src/hugr/validate.rs

```diff
@@ -159,14 +161,31 @@ impl<'a, 'b> ValidationContext<'a, 'b> {
             }
         }
 
-        // Check operation-specific constraints. Firstly that type args are correct
-        // (Good to call `resolve_extension_ops` immediately before this
-        //   - see https://github.com/CQCL-DEV/hugr/issues/508 )
+        // Check operation-specific constraints.
         if let OpType::LeafOp(crate::ops::LeafOp::CustomOp(b)) = op_type {
+            // Check TypeArgs are valid (in themselves, not necessarily wrt the TypeParams)
             for arg in b.args() {
                 arg.validate(self.extension_registry)
                     .map_err(|cause| ValidationError::SignatureError { node, cause })?;
             }
+            // Try to resolve serialized names to actual OpDefs in Extensions.
+            let e: Option<ExtensionOp>;
+            let ext_op = match &**b {
+                ExternalOp::Opaque(op) => {
+                    // If resolve_extension_ops has been called first, this would always return Ok(None)
+                    e = resolve_opaque_op(node, op, self.extension_registry)?;
+                    e.as_ref()
+                }
+                ExternalOp::Extension(ext) => Some(ext),
+            };
+            // If successful, check TypeArgs are valid for the declared TypeParams
+            if let Some(ext_op) = ext_op {
+                // WTF looks like check_opaque might be duplicating some of this?!
```

Comment:
Actually I think I can just drop this comment and maybe sort that out in another PR.